add a benchmark and profile script and hook into CI

danmayer opened the pull request:
This will make it easier for us to start upstreaming some of the performance and latency related PRs we have been testing out. We should be able to compare profiles before and after applying change and post any benchmark differences as well...

current benchmark:
```
❯❯❯$ bundle exec bin/benchmark                                                                                        <bundler> [main]
yjit: false
ruby 3.3.0 (2023-12-25 revision 5124f9ac75) [arm64-darwin23]
Warming up --------------------------------------
          client set   691.000 i/100ms
        raw sock set     1.403k i/100ms
Calculating -------------------------------------
          client set      6.391k (± 9.7%) i/s -     63.572k in  10.035280s
        raw sock set     14.013k (± 2.0%) i/s -    140.300k in  10.016703s

Comparison:
        raw sock set:    14012.6 i/s
          client set:     6391.3 i/s - 2.19x  slower
```

while a profile can be view with https://vernier.prof/

<img width="1920" alt="Screenshot 2025-02-10 at 10 00 39 PM (2)" src="https://github.com/user-attachments/assets/05d2ac6e-3f0f-48c0-bd4c-eb20217188f8" />

grcooper commented on bin/profile:
```diff
@@ -0,0 +1,190 @@
+    client.get(dalli_key) while Time.now - start_time < bench_time
+  end
+
+  Vernier.profile(out: 'socket_get_profile.json') do
```
Do we need raw socket profiles?

danmayer replied:
they were helpful in terms of seeing differences that were outside of the IO

danmayer commented on bin/benchmark:
```diff
@@ -0,0 +1,233 @@
+sock.setsockopt(Socket::SOL_SOCKET, Socket::SO_KEEPALIVE, true)
+# Benchmarks didn't see any performance gains from increasing the SO_RCVBUF buffer size
+# sock.setsockopt(Socket::SOL_SOCKET, ::Socket::SO_RCVBUF, 1024 * 1024 * 8)
+# Benchamrks did see an improvement in performance when increasing the SO_SNDBUF buffer size
```
OK, yeah dalli can also take these adjustments, but you are correct we should default to the same and only apply if folks are passing in options to adjust dalli and then also adjust the socket.

danmayer replied:
for now just dropping setting this, but as we look at tweaking better defaults we will want to try a few things out

danmayer commented on bin/benchmark:
```diff
@@ -0,0 +1,233 @@
+sock.flush
+sock.readline # clear the buffer
+
+# ensure we have basic data for the benchmarks and get calls
```
so these are the defaults for the get_multi... I can make the payload adjustable but we don't typically see get multi with 1mb values so I picked something that was more in the normal range... I will make it configurable and just have it be 1/10th of the full get / set size by default.

danmayer commented on bin/benchmark:
```diff
@@ -0,0 +1,233 @@
+    x.report('get sock') do
+      sock.write("get sock_key\r\n")
+      sock.readline
+      sock.read(payload.bytesize)
```
ok added an assert on all and we were missing a terminator read previously, this check caught that.

danmayer commented on bin/benchmark:
```diff
@@ -0,0 +1,233 @@
+TERMINATOR = "\r\n"
+puts "yjit: #{RubyVM::YJIT.enabled?}"
+
+client = Dalli::Client.new('localhost', serializer: StringSerializer, compress: false, raw: true)
```
yeah true, I will be getting this back to usage in CI soon once I port the set_multi feature.

grcooper commented on bin/benchmark:
```diff
@@ -0,0 +1,233 @@
+dalli_url = ENV['BENCH_CACHE_URL'] || '127.0.0.1:11211'
+
+ENV['BENCH_CACHE_URL'].gsub('unix://', '') if dalli_url.include?('unix')
+bench_target = ENV['BENCH_TARGET'] || 'set'
```
Fair enough, seems fine 👍

danmayer replied:
ok, doesn't seem important give one can call with anything, but moving both profile and bench to default to all

danmayer commented on .github/workflows/profile.yml:
```diff
@@ -0,0 +1,38 @@
+name: Profiles
+
+on: [push, pull_request]
```
I think it is good to have on any PR as we can review the profile if we have any concerns that it might impact performance

danmayer commented on .github/workflows/benchmarks.yml:
```diff
@@ -0,0 +1,26 @@
+    - name: Set up Ruby
+      uses: ruby/setup-ruby@v1
+      with:
+        ruby-version: 3.2
```
I set it at the current min version we support but we could set it at the latest as that is what we normally target

danmayer replied:
updated

danmayer commented on bin/benchmark:
```diff
@@ -0,0 +1,255 @@
+        count -= 1
+        tail = count.zero? ? '' : 'q'
+        sock.write(String.new("ms #{key} #{value_bytesize} c F0 T#{ttl} MS #{tail}\r\n",
+                              capacity: key.size + value_bytesize + 40) << value << TERMINATOR)
```
just needed to cover all the characters like ms, ' ', c, F0, etc.... I just picked a number as I was modifying this command a few times, a few extra unused bytes in the buffer didn't matter.

danmayer commented on bin/benchmark:
```diff
@@ -0,0 +1,255 @@
+  pairs.each do |key, value|
+    client.set(key, value, 3600, raw: true)
+  end
+end
```
the benchmark verifies during the get bench that they are their or it will raise an exception. So I don't think we need to check before the script, it would fail faster this way but also duplicate more code.

danmayer replied:
so handled with `raise 'mismatch' unless result == payload`

danmayer commented:
> I'm just getting around to reviewing this now after it's been merged but in general I find the code in bin/benchmark and bin/profile hard to parse and could be DRY'd up.

I will open a PR to fix up some more of this, but it is a bit hard to fully DRY up as I didn't want to create classes or other files just to leverage in the benchmark and profile, so there is definitely duplication between the two. Like tests, I often find fully drying up profiles and benchmarks makes it harder to focus in and change only the piece you care about digging into, so I don't want to get to fancy with abstractions that might leak and impact the code we are trying to examine.

danmayer commented on .github/workflows/profile.yml:
```diff
@@ -0,0 +1,38 @@
+    - name: Run Profiles
+      run: RUBY_YJIT_ENABLE=1 BENCH_TARGET=all bundle exec bin/profile
+    - name: Upload profile results
+      uses: actions/upload-artifact@v4
```
good call added documentation on the action file for how to get them

danmayer commented on bin/benchmark:
```diff
@@ -0,0 +1,255 @@
+  gem 'logger'
+end
+
+require_relative '../lib/dalli'
```
grouped all the others but the bundler one has to be above `gemfile do` block and others need to be below

danmayer commented on bin/benchmark:
```diff
@@ -0,0 +1,255 @@
+##
+# StringSerializer is a serializer that avoids the overhead of Marshal or JSON.
+##
+class StringSerializer
```
yup, updated

danmayer commented:
@nickamorim since this was closed and my other PR already was updating profile/benchmark to support meta protocol I addressed your feedback as part of that PR https://github.com/petergoldstein/dalli/pull/1029
